hummingbot/connector/exchange/bitmart/bitmart_exchange.py

```diff
@@ -205,7 +205,7 @@ async def _place_order(self,
 
     async def _place_cancel(self, order_id: str, tracked_order: InFlightOrder):
         api_params = {
```

Comment:
Symbol is required in V3 cancel order.

[Cancel Order V3 API Doc](https://developer-pro.bitmart.com/en/spot/#cancel-order-v3-signed)

`    async def _place_cancel(self, order_id: str, tracked_order: InFlightOrder):
        api_params = {
            "symbol": await self.exchange_symbol_associated_to_pair(tracked_order.trading_pair),
            "client_order_id": order_id,
        }`

Reply:
Ah yes, I just added it to the method call as a parameter similar to how place order makes use of the trading_pair. I'm not sure what upstream call will need to modify its params to pass in trading_pair but I have updated the call @nikspz

Reply:
Hey guys, just another community  member. I have cloned `babcoccl:fix/bitmart_api_update2` and tried to run locally but I am seeing lots of errors when the POST requests for placing orders are made


![image](https://github.com/user-attachments/assets/4d2ac965-f47d-4198-950c-fc21a02d407e)

Reply:
Good catch, I forgot to change the parameter 'params' to 'data' when I changed the methods to POST verbs from GETs. Fixed in the latest commit.

Reply:
Okay there's progress but still seeing issues on my end.



![image](https://github.com/user-attachments/assets/8c976136-7885-448f-a7dd-f9f693df42f6)

after I enter "stop" the bot enters into this state and basically stalls 
![image](https://github.com/user-attachments/assets/40feaf6f-0089-4dae-a4a0-8d18633fc233)

Reply:
> Hey guys, just another community member. I have cloned `babcoccl:fix/bitmart_api_update2` and tried to run locally but I am seeing lots of errors when the POST requests for placing orders are made
> 
> ![image](https://private-user-images.githubusercontent.com/32219869/391238929-4d2ac965-f47d-4198-950c-fc21a02d407e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI5Nzc3MjcsIm5iZiI6MTczMjk3NzQyNywicGF0aCI6Ii8zMjIxOTg2OS8zOTEyMzg5MjktNGQyYWM5NjUtZjQ3ZC00MTk4LTk1MGMtZmMyMWEwMmQ0MDdlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTMwVDE0MzcwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQwNmQxNjVkMDQxMmU2ZWRkMmIxZDlmNjgwYzc3YTNmZThjY2QyZWQ3YTU2NDc0ZjdlOGE3NDg0ZWY2ZTEzODImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.nWzxty9zWMo4_cq7o9KzsmM-_WGoVbcdrjeuE0fRt7E)

I think this is a bug in the connector itself and it's complaining about trying to loop through orders expecting an integer length but instead getting a string. I would guess that perhaps the broken API call is now working and returning a list that isn't in the format the codebase was expecting? Perhaps @encrypt3dw4lrus or @nikspz has some insight.

Reply:
@lockchainco in your log files for HB, is there a stack trace that points to a code line associated with these errors?

Reply:
I'm a bit tired as I've been trying to figure this out, will post my files with comments but I have some handlers added to client order tracker for now in order to not spam the console with failed to fetch status messages.

I did a lot of logging and their state is changing over to order_state 

`{
    "data":[
        {
            "symbol":"BTC_USDT",
            "side":"buy",
            "type":"market",
            "notional":"",
            "size":"1.0000000000",
            "ms_t":"1609926028000",
            "price":"46100.0000000000",
            "filled_notional":"46100.0000000000",
            "filled_size":"1.0000000000",
            "margin_trading":"0",
            "state":"4",
            "order_id":"2147857398",
            "order_type":"0",
            "last_fill_time":"1609926039226",
            "last_fill_price":"46100.00000",
            "last_fill_count":"1.00000",
            "exec_type":"M",
            "detail_id":"256348632",
            "client_order_id":"order4872191",
            "create_time":"1609926028000",
            "update_time":"1609926044000",
            "order_mode":"0",
            "entrust_type":"normal",
            "order_state":"partially_filled",
            "dealFee":"10.00"
        }
    ],
    "table":"spot/user/order"
}`

Order state (The field will be removed, please use the order_state field)

[Refer to this part search for order_state](https://developer-pro.bitmart.com/en/spot/#private-order-progress)

It still reads a state for me in some places so I added a fallback into my constants 

`ORDER_STATE = {
    "new": OrderState.OPEN,
    "partially_filled": OrderState.PARTIALLY_FILLED,
    "filled": OrderState.FILLED,
    "canceled": OrderState.CANCELED,
    "partially_canceled": OrderState.CANCELED,
    "-4": OrderState.OPEN,
    "-5": OrderState.PARTIALLY_FILLED,
    "-6": OrderState.FILLED,
    "-8": OrderState.CANCELED,
    "-12": OrderState.CANCELED,
}`


In `client_order_tracker.py` I added the following for now:

`async def _process_order_update(self, order_update: OrderUpdate):
        if not order_update or (not order_update.client_order_id and not order_update.exchange_order_id):
            self.logger().error(f"Invalid OrderUpdate: {order_update}", exc_info=True)
            return

        # Fetch tracked or lost orders
        tracked_order: Optional[InFlightOrder] = self.fetch_order(
            order_update.client_order_id, order_update.exchange_order_id
        )

        if tracked_order:
            previous_state = tracked_order.current_state
            updated = tracked_order.update_with_order_update(order_update)

            if updated:
                if order_update.new_state == OrderState.FILLED and not tracked_order.is_done:
                    try:
                        await asyncio.wait_for(
                            tracked_order.wait_until_completely_filled(), timeout=self.TRADE_FILLS_WAIT_TIMEOUT
                        )
                    except asyncio.TimeoutError:
                        self.logger().warning(
                            f"Timeout waiting for complete fills on {tracked_order.client_order_id}. "
                            f"Processing with available information."
                        )
                self._trigger_order_creation(tracked_order, previous_state, order_update.new_state)
                self._trigger_order_completion(tracked_order, order_update)
        else:
            # Handle untracked or lost orders
            lost_order = self.fetch_lost_order(
                client_order_id=order_update.client_order_id, exchange_order_id=order_update.exchange_order_id
            )
            if lost_order:
                if order_update.new_state in [OrderState.CANCELED, OrderState.FILLED, OrderState.FAILED]:
                    self.logger().info(f"Finalizing lost order {lost_order.client_order_id}.")
                    del self._lost_orders[lost_order.client_order_id]
            else:
                self.logger().debug(f"Untracked OrderUpdate: {order_update}")`
                
In `bitmart_exchange.py` I made many changes, it's getting late so I'll just paste in the full code:

`class BitmartExchange(ExchangePyBase):
    """
    BitmartExchange connects with BitMart exchange and provides order book pricing, user account tracking and
    trading functionality.
    """
    API_CALL_TIMEOUT = 10.0
    POLL_INTERVAL = 1.0
    UPDATE_ORDER_STATUS_MIN_INTERVAL = 10.0
    UPDATE_TRADE_STATUS_MIN_INTERVAL = 10.0

    web_utils = web_utils

    def __init__(self,
                 client_config_map: "ClientConfigAdapter",
                 bitmart_api_key: str,
                 bitmart_secret_key: str,
                 bitmart_memo: str,
                 trading_pairs: Optional[List[str]] = None,
                 trading_required: bool = True,
                 ):
        """
        :param bitmart_api_key: The API key to connect to private BitMart APIs.
        :param bitmart_secret_key: The API secret.
        :param trading_pairs: The market trading pairs which to track order book data.
        :param trading_required: Whether actual trading is needed.
        """
        self._api_key: str = bitmart_api_key
        self._secret_key: str = bitmart_secret_key
        self._memo: str = bitmart_memo
        self._trading_required = trading_required
        self._trading_pairs = trading_pairs

        super().__init__(client_config_map)
        self.real_time_balance_update = False

    @property
    def authenticator(self):
        return BitmartAuth(
            api_key=self._api_key,
            secret_key=self._secret_key,
            memo=self._memo,
            time_provider=self._time_synchronizer)

    @property
    def name(self) -> str:
        return "bitmart"

    @property
    def rate_limits_rules(self):
        return CONSTANTS.RATE_LIMITS

    @property
    def domain(self):
        return CONSTANTS.DEFAULT_DOMAIN

    @property
    def client_order_id_max_length(self):
        return CONSTANTS.MAX_ORDER_ID_LEN

    @property
    def client_order_id_prefix(self):
        return CONSTANTS.HBOT_ORDER_ID_PREFIX

    @property
    def trading_rules_request_path(self):
        return CONSTANTS.GET_TRADING_RULES_PATH_URL

    @property
    def trading_pairs_request_path(self):
        return CONSTANTS.GET_TRADING_RULES_PATH_URL

    @property
    def check_network_request_path(self):
        return CONSTANTS.CHECK_NETWORK_PATH_URL

    @property
    def trading_pairs(self):
        return self._trading_pairs

    @property
    def is_cancel_request_in_exchange_synchronous(self) -> bool:
        return True

    @property
    def is_trading_required(self) -> bool:
        return self._trading_required

    def supported_order_types(self) -> List[OrderType]:
        """
        :return a list of OrderType supported by this connector.
        Note that Market order type is no longer required and will not be used.
        """
        return [OrderType.LIMIT, OrderType.LIMIT_MAKER]

    def _is_request_exception_related_to_time_synchronizer(self, request_exception: Exception):
        error_description = str(request_exception)
        is_time_synchronizer_related = ("Header X-BM-TIMESTAMP" in error_description
                                        and ("30007" in error_description or "30008" in error_description))
        return is_time_synchronizer_related

    def _is_order_not_found_during_status_update_error(self, status_update_exception: Exception) -> bool:
        # TODO: implement this method correctly for the connector
        # The default implementation was added when the functionality to detect not found orders was introduced in the
        # ExchangePyBase class. Also fix the unit test test_lost_order_removed_if_not_found_during_order_status_update
        # when replacing the dummy implementation
        return False

    def _is_order_not_found_during_cancelation_error(self, cancelation_exception: Exception) -> bool:
        # TODO: implement this method correctly for the connector
        # The default implementation was added when the functionality to detect not found orders was introduced in the
        # ExchangePyBase class. Also fix the unit test test_cancel_order_not_found_in_the_exchange when replacing the
        # dummy implementation
        return False

    def _create_web_assistants_factory(self) -> WebAssistantsFactory:
        return web_utils.build_api_factory(
            throttler=self._throttler,
            time_synchronizer=self._time_synchronizer,
            auth=self._auth)

    def _create_order_book_data_source(self) -> OrderBookTrackerDataSource:
        return BitmartAPIOrderBookDataSource(
            trading_pairs=self._trading_pairs,
            connector=self,
            api_factory=self._web_assistants_factory)

    def _create_user_stream_data_source(self) -> UserStreamTrackerDataSource:
        return BitmartAPIUserStreamDataSource(
            auth=self._auth,
            trading_pairs=self._trading_pairs,
            connector=self,
            api_factory=self._web_assistants_factory,
        )

    def _get_fee(self,
                 base_currency: str,
                 quote_currency: str,
                 order_type: OrderType,
                 order_side: TradeType,
                 amount: Decimal,
                 price: Decimal = s_decimal_NaN,
                 is_maker: Optional[bool] = None) -> AddedToCostTradeFee:
        """
        To get trading fee, this function is simplified by using fee override configuration. Most parameters to this
        function are ignore except order_type. Use OrderType.LIMIT_MAKER to specify you want trading fee for
        maker order.
        """
        is_maker = order_type is OrderType.LIMIT_MAKER
        return AddedToCostTradeFee(percent=self.estimate_fee_pct(is_maker))

    async def _place_order(self,
                           order_id: str,
                           trading_pair: str,
                           amount: Decimal,
                           trade_type: TradeType,
                           order_type: OrderType,
                           price: Decimal,
                           **kwargs) -> Tuple[str, float]:

        api_params = {"symbol": await self.exchange_symbol_associated_to_pair(trading_pair),
                      "side": trade_type.name.lower(),
                      "type": "limit",
                      "size": f"{amount:f}",
                      "price": f"{price:f}",
                      "client_order_id": order_id,
                      }
        order_result = await self._api_post(
            path_url=CONSTANTS.CREATE_ORDER_PATH_URL,
            data=api_params,
            is_auth_required=True)
        exchange_order_id = str(order_result["data"]["order_id"])

        return exchange_order_id, self.current_timestamp

    async def _place_cancel(self, order_id: str, tracked_order: InFlightOrder) -> bool:
        """
        Sends a cancellation request to the exchange and returns whether the cancellation was successful.
        """
        try:
            api_params = {
                "symbol": await self.exchange_symbol_associated_to_pair(tracked_order.trading_pair),
                "client_order_id": order_id,
            }
            cancel_result = await self._api_post(
                path_url=CONSTANTS.CANCEL_ORDER_PATH_URL,
                data=api_params,
                is_auth_required=True,
            )
            return cancel_result.get("data", {}).get("result", False)
        except OSError as e:
            error_message = str(e)
            if "50031" in error_message: 
                self.logger().info(f"Order {order_id} is already completed. Marking as completed.")
                self._order_tracker.process_order_update(
                    OrderUpdate(
                        client_order_id=order_id,
                        exchange_order_id=tracked_order.exchange_order_id,
                        trading_pair=tracked_order.trading_pair,
                        update_timestamp=self.current_timestamp,
                        new_state=OrderState.FILLED,
                    )
                )
                return True
            raise

    async def _format_trading_rules(self, symbols_details: Dict[str, Any]) -> List[TradingRule]:
        """
        Converts json API response into a dictionary of trading rules.
        :param symbols_details: The json API response
        :return A dictionary of trading rules.
        Response Example:
        {
            "code": 1000,
            "trace":"886fb6ae-456b-4654-b4e0-d681ac05cea1",
            "message": "OK",
            "data": {
                "symbols": [
                    {
                        "symbol":"GXC_BTC",
                         "symbol_id":1024,
                         "base_currency":"GXC",
                         "quote_currency":"BTC",
                         "quote_increment":"1.00000000",
                         "base_min_size":"1.00000000",
                         "price_min_precision":6,
                         "price_max_precision":8,
                         "expiration":"NA",
                         "min_buy_amount":"0.00010000",
                         "min_sell_amount":"0.00010000"
                    },
                    ...
                ]
            }
        }
        """
        result = []
        for rule in symbols_details["data"]["symbols"]:
            if bitmart_utils.is_exchange_information_valid(rule):
                try:
                    trading_pair = await self.trading_pair_associated_to_exchange_symbol(rule["symbol"])
                    price_decimals = Decimal(str(rule["price_max_precision"]))
                    # E.g. a price decimal of 2 means 0.01 incremental.
                    price_step = Decimal("1") / Decimal(str(math.pow(10, price_decimals)))
                    result.append(TradingRule(trading_pair=trading_pair,
                                              min_order_size=Decimal(str(rule["base_min_size"])),
                                              min_order_value=Decimal(str(rule["min_buy_amount"])),
                                              min_base_amount_increment=Decimal(str(rule["base_min_size"])),
                                              min_price_increment=price_step))
                except Exception:
                    self.logger().exception(f"Error parsing the trading pair rule {rule}. Skipping.")
        return result

    async def _update_trading_fees(self):
        """
        Update fees information from the exchange
        """
        pass

    async def _update_balances(self):
        """
        Calls REST API to update total and available balances.
        """
        local_asset_names = set(self._account_balances.keys())
        remote_asset_names = set()
        account_info = await self._api_get(
            path_url=CONSTANTS.GET_ACCOUNT_SUMMARY_PATH_URL,
            is_auth_required=True)
        for account in account_info["data"]["wallet"]:
            asset_name = account["id"]
            self._account_available_balances[asset_name] = Decimal(str(account["available"]))
            self._account_balances[asset_name] = Decimal(str(account["available"])) + Decimal(str(account["frozen"]))
            remote_asset_names.add(asset_name)

        asset_names_to_remove = local_asset_names.difference(remote_asset_names)
        for asset_name in asset_names_to_remove:
            del self._account_available_balances[asset_name]
            del self._account_balances[asset_name]

    async def _request_order_update(self, order: InFlightOrder) -> Dict[str, Any]:
        try:
            response = await self._api_post(
                path_url=CONSTANTS.GET_ORDER_DETAIL_PATH_URL,
                data={"orderId": order.exchange_order_id},
                is_auth_required=True
            )
            if not response or "data" not in response:
                raise ValueError(f"Order update response is invalid or empty: {response}")
            return response
        except Exception as e:
            self.logger().error(f"Error fetching order update for {order.client_order_id}: {e}")
            raise


    async def _request_order_fills(self, order: InFlightOrder) -> Dict[str, Any]:
        return await self._api_post(
            path_url=CONSTANTS.GET_TRADE_DETAIL_PATH_URL,
            data={"symbol": await self.exchange_symbol_associated_to_pair(order.trading_pair)},
            is_auth_required=True)

    async def _all_trade_updates_for_order(self, order: InFlightOrder) -> List[TradeUpdate]:
        trade_updates = []

        try:
            if order.exchange_order_id is not None:
                all_fills_response = await self._request_order_fills(order=order)
                updates = self._create_order_fill_updates(order=order, fill_update=all_fills_response)
                trade_updates.extend(updates)
        except asyncio.CancelledError:
            raise
        except Exception as ex:
            is_error_caused_by_unexistent_order = '"code":50005' in str(ex)
            if not is_error_caused_by_unexistent_order:
                raise

        return trade_updates

    async def _request_order_status(self, tracked_order: InFlightOrder) -> Optional[OrderUpdate]:
        try:
            updated_order_data = await self._request_order_update(order=tracked_order)

            if "data" not in updated_order_data:
                self.logger().warning(f"Missing 'data' in response for order {tracked_order.client_order_id}: {updated_order_data}")
                return None

            order_data = updated_order_data["data"]

            state_field = order_data.get("order_state", order_data.get("state"))
            if state_field is None:
                self.logger().warning(f"Neither 'order_state' nor 'state' found in order data: {order_data}")
                return None

            new_state = CONSTANTS.ORDER_STATE.get(str(state_field), OrderState.UNKNOWN)
            if new_state == OrderState.UNKNOWN:
                self.logger().warning(f"Unknown state encountered: {state_field} for order {tracked_order.client_order_id}")

            return OrderUpdate(
                client_order_id=tracked_order.client_order_id,
                exchange_order_id=str(order_data.get("order_id", "UNKNOWN")),
                trading_pair=tracked_order.trading_pair,
                update_timestamp=self.current_timestamp,
                new_state=new_state,
            )
        except Exception as e:
            self.logger().error(f"Error fetching status for order {tracked_order.client_order_id}: {e}")
            return None

    def _create_order_fill_updates(self, order: InFlightOrder, fill_update: Dict[str, Any]) -> List[TradeUpdate]:
        updates = []
        try:
            # Handle when "data" is a list
            if isinstance(fill_update["data"], list):
                fills_data = fill_update["data"]
            elif "trades" in fill_update["data"]:
                fills_data = fill_update["data"]["trades"]
            else:
                raise KeyError("'data' or 'trades' not found in fill_update")
        except Exception as e:
            self.logger().error(f"Error in fill_update structure: {e}, fill_update: {fill_update}")
            return updates

        for fill_data in fills_data:
            fee = TradeFeeBase.new_spot_fee(
                fee_schema=self.trade_fee_schema(),
                trade_type=order.trade_type,
                percent_token=fill_data.get("feeCoinName", ""),
                flat_fees=[TokenAmount(amount=Decimal(fill_data.get("fee", "0")), token=fill_data.get("feeCoinName", ""))]
            )
            trade_update = TradeUpdate(
                trade_id=str(fill_data.get("tradeId")),
                client_order_id=order.client_order_id,
                exchange_order_id=str(fill_data.get("orderId")),
                trading_pair=order.trading_pair,
                fee=fee,
                fill_base_amount=Decimal(fill_data.get("size", "0")),
                fill_quote_amount=Decimal(fill_data.get("notional", "0")),
                fill_price=Decimal(fill_data.get("price", "0")),
                fill_timestamp=int(fill_data.get("createTime", 0)) * 1e-3,
            )
            updates.append(trade_update)

        return updates


    def _create_order_update(self, order: InFlightOrder, order_update: Dict[str, Any]) -> OrderUpdate:
        order_data = order_update["data"]
        new_state = CONSTANTS.ORDER_STATE[order_data["status"]]
        update = OrderUpdate(
            client_order_id=order.client_order_id,
            exchange_order_id=str(order_data["order_id"]),
            trading_pair=order.trading_pair,
            update_timestamp=self.current_timestamp,
            new_state=new_state,
        )
        return update

    async def _user_stream_event_listener(self):
        async for event_message in self._iter_user_event_queue():
            try:
                event_type = event_message.get("table")
                execution_data = event_message.get("data", [])

                if event_type == CONSTANTS.PRIVATE_ORDER_PROGRESS_CHANNEL_NAME:
                    for each_event in execution_data:
                        try:
                            client_order_id: Optional[str] = each_event.get("client_order_id")
                            fillable_order = self._order_tracker.all_fillable_orders.get(client_order_id)
                            updatable_order = self._order_tracker.all_updatable_orders.get(client_order_id)

                            # Prefer "order_state" but fallback to "state"
                            state_field = each_event.get("order_state", each_event.get("state"))
                            if state_field is None:
                                self.logger().warning(f"State field missing in event: {each_event}")
                                continue

                            try:
                                new_state = CONSTANTS.ORDER_STATE[str(state_field)]
                            except KeyError:
                                self.logger().warning(f"Unknown state encountered: {state_field} for event {each_event}")
                                new_state = OrderState.UNKNOWN

                            # Process the order update based on the new_state
                            if updatable_order is not None:
                                order_update = OrderUpdate(
                                    trading_pair=updatable_order.trading_pair,
                                    update_timestamp=self.current_timestamp,
                                    new_state=new_state,
                                    client_order_id=client_order_id,
                                    exchange_order_id=each_event.get("order_id"),
                                )
                                self._order_tracker.process_order_update(order_update=order_update)

                        except Exception:
                            self.logger().exception("Unexpected error processing user stream event.")
            except Exception:
                self.logger().exception("Unexpected error in user stream listener loop.")

    def _initialize_trading_pair_symbols_from_exchange_info(self, exchange_info: Dict[str, Any]):
        mapping = bidict()
        for symbol_data in filter(bitmart_utils.is_exchange_information_valid, exchange_info["data"]["symbols"]):
            mapping[symbol_data["symbol"]] = combine_to_hb_trading_pair(base=symbol_data["base_currency"],
                                                                        quote=symbol_data["quote_currency"])
        self._set_trading_pair_symbol_map(mapping)

    async def _get_last_traded_price(self, trading_pair: str) -> float:
        params = {
            "symbol": await self.exchange_symbol_associated_to_pair(trading_pair=trading_pair)
        }

        resp_json = await self._api_get(
            path_url=CONSTANTS.GET_LAST_TRADING_PRICES_PATH_URL,
            params=params
        )

        return float(resp_json["data"]["tickers"][0]["last_price"])`

I'll continue diving more into this tomorrow and hopefully can get the statuses to work, I know some of this stuff is temporary fixes but it does trade, cancel successfully, handles order fills so it's on the right track. 

![image](https://github.com/user-attachments/assets/0891c058-e9de-49c3-9c3f-e94009d97e89)
![image](https://github.com/user-attachments/assets/affec04c-f8f6-479b-be53-7f72899b020b)

Reply:
@encrypt3dw4lrus Good work! So you're seeing some messages coming back with a "state" field and others coming back with the new "order_state" field? That seems odd, I would have thought just changing line 377 to handle the new "order_state" name would have worked. Thanks for the good work!

Reply:
I hit a bit of a brick wall today got all the updates fetching etc. but even in my last reply you can see 
![image](https://github.com/user-attachments/assets/82a6315d-0650-4ab6-a055-222b0367875d)
The filled amount keeps stacking, this skewes the available balance and eventually you end up with only bids or asks.

Reply:
I figured this was caused by orders duplicating and the cumulative filled stacking due to this. I will leave it to run for 5-6 hours and check if I mitigated it by skipping duplicates in the `_create_order_fill_updates` and making few changes to `update_with_trade_update`

I am working on this stuff for the first time so doing my best, maybe not the best approach, but if it works correctly not skewing my balance then it's a success and can further build from there. I was ending up with like -1273 USDT available balance, USDC available at 798 even though I only got 400$ for testing on here...So let's see how the results pan out after good few hours of testing. Will post an update in 8-12 hours.

Reply:
Okay so it works as it should, it doesn't keep track of PNL and in telegram won't show trade history but at least it does what it is meant to do, hang orders as it should, keeps track of available balance correctly. I might fork this repo PR and make my own branch for it just so someone can copy my code and then we can work up from there to keep track of pnl, instead of skipping duplicates, work on avoiding duplicates completely and hopefully we'll get this to work fully as it should as we build on it gradually.

Reply:
@encrypt3dw4lrus really good work and I'm completely new to this domain as well so no judgement from me. Definitely create your own fork and throw your code changes on it. I would say if it's trading correctly submit the PR and we can tackle PNL and the other issue separately.